File: demos/tftpaint/python/tftpaint.py
import sys
import time
import platform

import threading
import queue

import serial
import serial.tools.list_ports

import tkinter as tk

if platform.system() == 'Windows':      # Windows - COM*
    serialPort = 'COM5'
elif platform.system() == 'Darwin':     # macOS - /dev/tty.usbmodem* ou /dev/tty.usbserial*
    serialPort = '/dev/tty.usbmodem22101'
else:                                   # Linux - /dev/ttyUSB* ou /dev/ttyACM*
    serialPort = '/dev/ttyACM0'

if not serialPort in [x.device for x in list(serial.tools.list_ports.comports())]:
    print('Error: Serial port {} not available.'.format(serialPort))
    sys.exit()

colors888 = ['#F800F8', '#0000F8', '#00FCF8', '#00FC00', '#F8FC00', '#F80000']

currentColorIndex = 5

stopThread = False

queueCom = queue.Queue()

threadCom = None

incomingMsg = ''

def read_from_port():

    global incomingMsg

    while True:

        if stopThread:
            break

        n = ser.in_waiting

        if n > 0:

            reading = ser.read(1)

            incomingMsg += reading.decode('utf-8')

            if incomingMsg[-1] == '\n':
                queueCom.put(incomingMsg)
                incomingMsg = ''

        else:
            time.sleep(0.01)


def periodicCall():

    global currentColorIndex

    while queueCom.qsize():

        msg = queueCom.get(0).rstrip()

        if msg:

            if msg[0]== 'X':
                xy = [int(x) for x in msg.split('X')[1].split('Y')]
                drawPoint(xy[1], xy[0], colors888[currentColorIndex])
            elif msg[0]== '!':
                canvas.delete('point')
            elif msg[0]== 'R':
                currentColorIndex = 5
            elif msg[0]== 'Y':
                currentColorIndex = 4
            elif msg[0]== 'G':
                currentColorIndex = 3
            elif msg[0]== 'C':
                currentColorIndex = 2
            elif msg[0]== 'B':
                currentColorIndex = 1
            elif msg[0]== 'M':
                currentColorIndex = 0

    root.after(5, periodicCall)


def drawPoint(x, y, c):

    r = 3
    y = 240-y
    canvas.create_oval(x-r, y-r, x+r, y+r, fill=c, outline=c, tag='point')
    pass


def quit():

    global stopThread

    ser.close()
    root.quit()

    if threadCom.is_alive():
        stopThread = True
        threadCom.join()


root = tk.Tk()

root.geometry('320x240')

canvas = tk.Canvas(root, background='black')
canvas.place(x=0, y=0, width=320, height=240)

canvas.create_rectangle(0, 0, 40, 240, outline='white')

for i in range(0,6):
    canvas.create_rectangle(41, i*40, 2*40, (i+1)*40, fill=colors888[i], outline=colors888[i])

root.protocol("WM_DELETE_WINDOW", quit)

ser = serial.Serial(port=serialPort, baudrate=9600)
time.sleep(3) # Chargement du bootloader par l'Arduino

threadCom = threading.Thread(target=read_from_port)
threadCom.start()

root.after(5, periodicCall)

root.mainloop()

File: demos/tftpaint/matlab/tftpaint.m
function tftpaint

if ispc         % Windows - COM*
    serialPort = 'COM9';
elseif ismac	% macOS - /dev/tty.usbmodem* ou /dev/tty.usbserial*
    serialPort = '/dev/tty.usbmodem21101';
else            % Linux - /dev/ttyUSB* ou /dev/ttyACM*
    serialPort = '/dev/ttyACM0';
end

freeports = serialportlist("available");
if ~any(contains(freeports, serialPort))
    error('Serial port %s not available.', serialPort)
end

quitFunction = false;

colors888 = [0.9725 0 0
    0.9725 0.9882 0
    0 0.9882 0
    0 0.9882 0.9725
    0 0 0.9725
    0.9725 0 0.9725];

currentColorIndex = 1;

h_figure = figure('units', 'pixels', ...
    'position', [0 0 320 240], ...
    'menubar', 'none', ...
    'toolbar', 'none', ...    'renderer', 'painter', ...
    'closerequestfcn', @quit);

axes('units', 'pixels', ...
    'position', [0 0 320 240], ...
    'color', 'black', ...
    'box', 'off', ...
    'xlim', [0 320], ...
    'ylim', [0 240], ...
    'xtick', [], ...
    'ytick', [], ...
    'xcolor', 'white', ...
    'ycolor', 'white', ...
    'xlimmode', 'manual', ...
    'ylimmode', 'manual', ...
    'nextplot', 'add', ...
    'DataAspectRatio', [1 1 1], ...
    'PlotBoxAspectRatio', [3 4 4]);

line([1 40 40 1 ; 40 40 1 1], [1 1 240 240 ; 1 240 240 1], zeros(2,4), 'color', 'white', 'linewidth', 2)

for i = 1:6
    rectangle('position', [40 (i-1)*40 39 39], 'facecolor', colors888(i,:), 'edgecolor', colors888(i,:))
end

ser = serialport(serialPort, 9600);
set(h_figure, 'pointer', 'watch')
pause(3) % Chargement du bootloader par l'Arduino
set(h_figure, 'pointer', 'arrow')

    function readSerialData

        if ser.NumBytesAvailable

            str = readline(ser);

            if str{1}(1) == 'X'
                xy = sscanf(str, 'X%dY%d');
                drawPoint(xy(2), xy(1), colors888(currentColorIndex,:));
            elseif str{1}(1) == '!'
                h = findobj('tag', 'point');
                delete(h)
            elseif str{1}(1) == 'R'
                currentColorIndex = 1;
            elseif str{1}(1) == 'Y'
                currentColorIndex = 2;
            elseif str{1}(1) == 'G'
                currentColorIndex = 3;
            elseif str{1}(1) == 'C'
                currentColorIndex = 4;
            elseif str{1}(1) == 'B'
                currentColorIndex = 5;
            elseif str{1}(1) == 'M'
                currentColorIndex = 6;
            end

        end

    end

    function quit(~, ~)

        delete(ser)

        delete(gcf)
        quitFunction = true;

    end

    function drawPoint(x, y, color)
        plot(x, y, ...
            'color', color, ...
            'marker', 'o', ...
            'markerfacecolor', color, ...
            'linestyle', 'none', ...
            'markersize', 4, ...
            'tag', 'point')
    end

while 1

    drawnow;

    if quitFunction
        break;
    end

    readSerialData

end

end